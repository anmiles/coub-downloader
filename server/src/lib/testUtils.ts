import type { Coub } from 'types/coub';

function createTestCoub(id: string, override?: Partial<Coub>): Coub {
	return {
		title         : `Title: ${id}`,
		permalink     : id,
		picture       : `${id}.jpg`,
		likes_count   : 1,
		views_count   : 1,
		file_versions : {
			html5 : {
				video : {},
				audio : {},
			},
		},
		media_blocks : {
			external_raw_videos : [],
		},
		...override,
	};
}

export { createTestCoub };
export default { createTestCoub };
