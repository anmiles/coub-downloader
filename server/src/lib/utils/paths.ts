import path from 'path';

import type { templates } from '../renderer';

const dirPaths = {
	input    : 'input',
	output   : 'output',
	// TODO: Remove this hack after moving to React
	templates: path.relative(process.cwd(), path.join(__dirname, '../../templates')),
};

export function getCoubsFile(profile: string): string {
	return path.join(dirPaths.input, `${profile}.json`);
}

export function getProfileDir(profile: string): string {
	return path.join(dirPaths.output, profile);
}

export function getIndexFile(profile: string): string {
	return path.join(getProfileDir(profile), 'index.html');
}

export function getMediaDir(profile: string): string {
	return path.join(getProfileDir(profile), 'media');
}

export function getMediaFile(profile: string, coubID: string, mediaURL: string): string {
	const ext = mediaURL.split('.').pop();
	return path.join(getMediaDir(profile), coubID, `${coubID}.${ext}`);
}

export function getProfilesFile(): string {
	return path.join(dirPaths.input, 'profiles.json');
}

export function getTemplateFile(templateName: keyof typeof templates): string {
	return path.join(dirPaths.templates, `${templateName}.html`);
}
