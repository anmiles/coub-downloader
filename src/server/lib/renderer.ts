import fs from 'fs';
import path from 'path';
import _ from 'lodash';
import type { Coub } from '../types/coub';

export default class Renderer {
	render(htmlFilename: string, templatePath: string, coubs: Coub[]): void {
		const templates = this.getTemplates(templatePath);
		const formattedCoubs = coubs.map(coub => this.formatCoub(coub, templates));
		const html = this.format('index', { coubs: formattedCoubs.join('\n') }, templates);
		fs.writeFileSync(htmlFilename, html);
	}

	getTemplates(templatePath: string): Record<string, string> {
		return fs.readdirSync(templatePath).reduce((obj, filename) => {
			const name = filename.split('.')[0];
			const html = fs.readFileSync(path.join(templatePath, filename)).toString().trim();
			obj[name] = html;
			return obj;
		}, {} as Record<string, string>);
	}

	formatCoub(coub: Coub, templates: Record<string, string>): string {
		this.preventUnsafeString(coub.permalink, 'coub ID');

		coub.title = this.escapeString(coub.title);

		if (!coub.file_versions.html5.audio || Object.keys(coub.file_versions.html5.audio).length === 0) {
			_.set(coub, 'audio', '&nbsp;');
		}

		if (coub.media_blocks.external_raw_videos.length === 0) {
			_.set(coub, 'externals', '&nbsp;');
		} else {
			_.set(coub, 'externals', coub.media_blocks.external_raw_videos.map(external => {
				this.preventUnsafeString(external.meta.service, 'service name of external video');
				external.title = this.escapeString(external.title);
				return this.format('external', external, templates);
			}).join('\n'));
		}

		if (typeof coub.likes_count !== 'number') throw `coub.likes_count is ${coub.likes_count} that is not a number`;
		if (typeof coub.views_count !== 'number') throw `coub.views_count is ${coub.views_count} that is not a number`;

		return this.format('coub', coub, templates);
	}

	preventUnsafeString(s: string, description: string): void {
		if (typeof s !== 'string') {
			throw `${description} is not string but ${typeof s} and might be unsafe: ${s}`;
		}

		if (!/^\w+$/.test(s)) {
			throw `${description} is not alphanumeric string and might be unsafe: ${s}`;
		}
	}

	escapeString(s: string): string {
		const lookup: Record<string, string> = {
			'&': '&amp;',
			'"': '&quot;',
			'<': '&lt;',
			'>': '&gt;'
		};

		return s.replace( /[&"<>]/g, (c: string) => lookup[c] );
	}

	format(rootTemplate: string, json: Record<string, unknown>, templates: Record<string, string>): string {
		const regex = /\{\{([^}]+)\}\}/g;
		let html = templates[rootTemplate];

		do {
			html = html.replace(regex, (match, key: string) => (_.get(json, key) || _.get(templates, key)) as string);
		} while (regex.test(html));

		return html;
	}
}
