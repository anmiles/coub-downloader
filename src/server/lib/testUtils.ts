export { createTestCoub };
import type { Coub } from '../types/coub';

function createTestCoub(id: string, override?: Partial<Coub>): Coub {
	return {
		permalink: id,
		title: `Title: ${id}`,
		file_versions: {
			html5: {
				video: {},
				audio: {}
			}
		},
		media_blocks: {
			external_raw_videos: [],
		},
		picture: `${id}.jpg`,
		likes_count: 1,
		views_count: 1,
		...override
	};
}
