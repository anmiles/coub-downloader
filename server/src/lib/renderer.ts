import fs from 'fs';

import type { Coub } from '@coub-downloader/shared';
import _ from 'lodash';

import { getTemplateFile } from './utils/paths';

export const templates = {
	index     : [ 'style', 'page', 'script' ] as const,
	page      : [ 'content' ] as const,
	style     : [ ] as const,
	script    : [ ] as const,
	coub      : [ 'video', 'audio', 'fullscreen', 'stats', 'title', 'externals' ] as const,
	video     : [ 'permalink' ] as const,
	audio     : [ 'permalink' ] as const,
	fullscreen: [ ] as const,
	stats     : [ 'likes', 'views' ] as const,
	external  : [ 'service', 'url', 'title' ] as const,
} as const;

type TemplateName = keyof typeof templates;

const allHTML = {} as Record<TemplateName, string>;	// eslint-disable-line @typescript-eslint/no-unsafe-type-assertion

export function renderIndex(coubs: Coub[]): string {
	const formattedCoubs = coubs.map(renderCoub);
	const content        = formattedCoubs.join('\n');

	const style  = render('style', {});
	const script = render('script', {});
	const page   = render('page', { content });
	return render('index', { style, page, script });
}

function renderCoub(coub: Coub): string {
	const permalink = ensureStringSafe(coub.permalink, 'Coub ID');
	const video     = render('video', { permalink });

	const audio = coub.file_versions.html5.audio && Object.keys(coub.file_versions.html5.audio).length > 0
		? render('audio', { permalink })
		: '';

	const fullscreen = render('fullscreen', {});

	const stats = render('stats', {
		likes: coub.likes_count.toString(),
		views: coub.views_count.toString(),
	});

	const title     = escapeString(coub.title);
	const externals = coub.media_blocks.external_raw_videos.length === 0
		? '&nbsp;'
		: coub.media_blocks.external_raw_videos.map((external) => {
				const service = ensureStringSafe(external.meta.service, 'Service name of external video');
				const url     = external.url;
				const title   = escapeString(external.title);
				return render('external', { service, url, title });
			}).join('\n');

	return render('coub', { video, audio, fullscreen, stats, title, externals });
}

function ensureStringSafe(s: unknown, description: string): string {
	if (typeof s !== 'string') {
		throw new Error(`${description} is not string but ${typeof s} and might be unsafe: ${String(s)}`);
	}

	if (!/^\w+$/.test(s)) {
		throw new Error(`${description} is not alphanumeric string and might be unsafe: ${s}`);
	}

	return s;
}

function escapeString(s: string | undefined): string | undefined {
	if (typeof s === 'undefined') { return undefined; }

	for (const { symbol, entity } of [
		{ symbol: '&', entity: '&amp;' },
		{ symbol: '"', entity: '&quot;' },
		{ symbol: '<', entity: '&lt;' },
		{ symbol: '>', entity: '&gt;' },
	]) {
		s = s.replaceAll(symbol, entity);
	}

	return s;
}

// TODO: Use react
function render<T extends TemplateName>(templateName: T, values: Record<typeof templates[T][number], string | undefined>): string {
	let html        = getTemplate(templateName);
	const allValues = values as Record<typeof templates[TemplateName][number], string | undefined>;

	for (const variable of templates[templateName]) {
		const value = allValues[variable];

		if (typeof value === 'undefined') {
			throw new Error(`Missing required value '${variable}' while rendering template '${templateName}'`);
		}

		html = html.replaceAll(`\${${variable}}`, value);
	}

	return html;
}

function getTemplate(templateName: TemplateName): string {
	if (!(templateName in allHTML)) {
		const file            = getTemplateFile(templateName);
		const template        = fs.readFileSync(file).toString().trim();
		allHTML[templateName] = template;
	}

	return allHTML[templateName];
}
