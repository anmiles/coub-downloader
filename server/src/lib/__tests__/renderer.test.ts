import { renderIndex } from '../renderer';

import { createTestCoub } from './utils/testUtils';

const internalCoub = createTestCoub('internalCoub');

const externalCoub = createTestCoub('externalCoub', {
	media_blocks: {
		external_raw_videos: [
			{ title: 'title1', url: 'url1', meta: { service: 'service1' } },
			{ title: 'title2', url: 'url2', meta: { service: 'service2' } },
		],
	},
});

const audioCoub = createTestCoub('audioCoub', {
	file_versions: {
		html5: {
			audio: {
				big: {
					size: 10,
					url : 'URL',
				},
			},
			video: {},
		},
	},
});

const coubs = [ internalCoub, externalCoub, audioCoub ];

describe('src/lib/renderer', () => {
	describe('renderIndex', () => {
		it('should return index page', () => {
			const result = renderIndex(coubs);

			expect(result).toMatchSnapshot();
		});

		it('should throw error if property missing', () => {
			const untitledCoub = createTestCoub('untitledCoub', { title: undefined });

			const func = (): string => renderIndex([ ...coubs, untitledCoub ]);

			expect(func).toThrow('Missing required value \'title\' while rendering template \'coub\'');
		});

		it('should throw error if string property is actually not a string', () => {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-type-assertion
			const numericCoub = createTestCoub('numericCoub', { permalink: 10 as unknown as string });

			const func = (): string => renderIndex([ ...coubs, numericCoub ]);

			expect(func).toThrow('Coub ID is not string but number and might be unsafe: 10');
		});

		it('should throw error if string property contains non-alphanumeric characters', () => {
			const hyphenCoub = createTestCoub('coub3', { permalink: 'a-b' });

			const func = (): string => renderIndex([ ...coubs, hyphenCoub ]);

			expect(func).toThrow('Coub ID is not alphanumeric string and might be unsafe: a-b');
		});
	});
});
