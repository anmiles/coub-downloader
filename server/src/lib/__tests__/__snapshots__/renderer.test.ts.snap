// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/lib/renderer renderIndex should return index page 1`] = `
"<!DOCTYPE html>
<html>
<head>
	<title>Coub</title>
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
</head>
<body>

<style type="text/css">
	body {
		font-family: Helvetica, Arial, "Lucida Grande", sans-serif;
		font-size: 15px;
	}

	.coub {
		width: 640px;
		max-width: 100%;
		margin: 2em auto;
		position: relative;
	}

	video {
		width: 100%;
		max-height: 480px;
		background-color: #fafafa;
		border: 1px solid #e1e1e1;
		border-radius: 6px;
		cursor: pointer;
	}

	audio {
		display: none;
	}

	.fullscreen-toggle {
		font-size: 2em;
		color: white;
		position: absolute;
		right: 0.25em;
		top: 0.25em;
		line-height: 100%;
		cursor: pointer;
		opacity: 0.5;
	}

	.fullscreen-toggle:hover {
		opacity: 0.75;
	}

	:fullscreen video {
		max-height: none;
		height: 100%;
		background: none;
		border: none;
	}

	.description {
		margin-top: 0.5em;
	}

	.description * {
		line-height: 26px;
	}

	.stats {
		float: right;
		margin-left: 1em;
	}

	.title {
		font-size: 22px;
		font-weight: bold;
		letter-spacing: -0.4px;
	}

	.external-link {
		color: #2f4ed9;
	}

	.like {
		color: red;
	}
</style>

<div class="coubs">

<div class="coub">
	<video loop muted preload="none" poster="media/internalCoub/internalCoub.jpg" src="media/internalCoub/internalCoub.mp4"></video>
	
	<a class="fullscreen-toggle">&#x26F6;</a>
	<div class="description">
		<div class="stats"><span class="like">&#x2764;</span> 1 &nbsp; &#x1F441; 2</div>
		<div class="title">Title: internalCoub</div>
		&nbsp;
	</div>
</div>
<div class="coub">
	<video loop muted preload="none" poster="media/externalCoub/externalCoub.jpg" src="media/externalCoub/externalCoub.mp4"></video>
	
	<a class="fullscreen-toggle">&#x26F6;</a>
	<div class="description">
		<div class="stats"><span class="like">&#x2764;</span> 1 &nbsp; &#x1F441; 2</div>
		<div class="title">Title: externalCoub</div>
		<div class="external">service1: <a class="external-link" href="url1" target="_blank">title1</a></div>
<div class="external">service2: <a class="external-link" href="url2" target="_blank">title2</a></div>
	</div>
</div>
<div class="coub">
	<video loop muted preload="none" poster="media/audioCoub/audioCoub.jpg" src="media/audioCoub/audioCoub.mp4"></video>
	<audio src="media/audioCoub/audioCoub.mp3" loop controls></audio>
	<a class="fullscreen-toggle">&#x26F6;</a>
	<div class="description">
		<div class="stats"><span class="like">&#x2764;</span> 1 &nbsp; &#x1F441; 2</div>
		<div class="title">Title: audioCoub</div>
		&nbsp;
	</div>
</div>

</div>

<script type="text/javascript">
window.addEventListener('load', function () {
	var currentVideo = null;
	var topEdges = [];

	function start(video) {
		if (currentVideo) {
			if (currentVideo === video) return;
			currentVideo.pause();
		}
		currentVideo = video;
		video.play();
	}

	function stop(video) {
		currentVideo = null;
		video.pause();
	}

	Array.from(document.querySelectorAll('video')).forEach(function(video) {
		const container = video.parentNode;
		video.muted = true;
		topEdges[container.offsetTop] = video;
		var audio = video.nextElementSibling;
		var fullscreen = audio.nextElementSibling;

		video.onplay = function(){
			audio.play();
		};

		video.onpause = function(){
			audio.pause();
		};

		video.onclick = function() {
			video === currentVideo ? stop(video) : start(video);
		};

		fullscreen.onclick = function() {
			document.fullscreenElement ? document.exitFullscreen() : container.requestFullscreen();
		};
	});

	var topEdgeKeys = Object.keys(topEdges);

	window.scrollTimeout = null;

	window.onscroll = function(){
		if (window.scrollTimeout) clearTimeout(window.scrollTimeout);

		window.scrollTimeout = setTimeout(function(){
			for (var i in topEdgeKeys) {
				if (topEdgeKeys[i] > window.scrollY) {
					if (currentVideo) {
						start(topEdges[topEdgeKeys[i]]);
					}
					break;
				}
			}
		}, 100);
	};
});
</script>

</body>
</html>"
`;
